[16.0][MIG] account_invoice_validation_queued

Saran440 opened the pull request:
Standard MIgrated and Update readme

pedrobaeza committed:
[ADD] account_invoice_validation_queued: Enqueue account invoice validation

This module allows to enqueue in several jobs the account validation process
to be executed in paralell on background, which is normally done serially and
on foreground.

Each invoice creates a job for performing the validation, but you will only be
able to enqueue invoices for the same date.

Installation
============

This module depends on *queue_job* module that is hosted on
https://github.com/OCA/queue.

Configuration
=============

Jobs are enqueued in the channel ``root.account_invoice_validation_queued``,
so you must adjust your
`Odoo configuration <https://github.com/OCA/queue/tree/11.0/queue_job#configuration>`_
according this.

If you want to see queued jobs, you need "Job Queue / Job Queue Manager"
permission in your user.

Usage
=====

* Go to *Invoicing > Sales > Documents > Customer Invoices* or
  *Invoicing > Purchases > Documents > Vendor Bills*.
* Mark at least one check on the left part of one draft invoice line in the
  list view.
* Click on *Action > Confirm Draft Invoice*.
* On the dialog popup that appears, click on "Enqueue Validation".
* If any of the selected invoices are not in draft state, you will receive
  an error.
* If any of the invoices is already enqueued, there will be a message saying
  so and avoiding to perform the process.
* Once enqueued, and having the "Job Queue Manager" permission, you can go to
  the invoice, and see the tab "Validation Jobs". A list with all the jobs
  related to that invoice can be found there.

Known issues / Roadmap
======================

* Allow to enqueue for validating invoices of different date.
* There's a chance that if you perform several enqueues of invoice validations
  from different dates, the order of the validated invoices, and thus, the
  number given for it, will result disordered.

pedrobaeza committed:
[IMP] *_queued: Remove `cancel` method + use same user

pedrobaeza committed:
[MIG] account_invoice_validation_queued: Migration to 12.0

* Standard procedure
* README updated
* Promoted to Production/Stable

pedrobaeza committed:
[IMP] : black, isort, prettier

Commit:
[MIG] account_invoice_validation_queued: Migration to 13.0

oca-travis committed:
[UPD] Update account_invoice_validation_queued.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_invoice_validation_queued
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_invoice_validation_queued/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_invoice_validation_queued
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_invoice_validation_queued/

pedrobaeza committed:
[IMP] account_invoice_validation_queued: Avoid m2m table collapse

Having other m2m pointing to queue.job will make them to share the same table and
get akward results.

We avoid it specifying a table. It includes migration script for moving data from
one table to the other.

pedrobaeza committed:
[FIX] account_invoice_validation_queued: Proper SQL column

anasuarez1 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (7 of 7 strings)

Translation: account-invoicing-13.0/account-invoicing-13.0-account_invoice_validation_queued
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_invoice_validation_queued/es/

iwkse committed:
[14.0][MIG] account_invoice_validation_queued

oca-travis committed:
[UPD] Update account_invoice_validation_queued.pot

OCA-git-bot committed:
[UPD] README.rst

ofonseca-pyming committed:
Added translation using Weblate (Catalan)

BT-jcolmeiro committed:
[IMP] account_invoice_validation_queued: black, isort, prettier

BT-jcolmeiro committed:
[15.0][MIG] account_invoice_validation_queued: Migration to 15.0

Commit:
[UPD] Update account_invoice_validation_queued.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-15.0/account-invoicing-15.0-account_invoice_validation_queued
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-15-0/account-invoicing-15-0-account_invoice_validation_queued/

OCA-git-bot committed:
[UPD] README.rst

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (7 of 7 strings)

Translation: account-invoicing-15.0/account-invoicing-15.0-account_invoice_validation_queued
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-15-0/account-invoicing-15-0-account_invoice_validation_queued/es/

Saran440 committed:
[IMP] account_invoice_validation_queued: pre-commit stuff

Saran440 committed:
[MIG] account_invoice_validation_queued: Migration to 16.0

hbrunn reviewed (approved):
/ocabot migration account_invoice_validation_queued

francesco-ooops commented:
@Saran440 (and @pedrobaeza as module author): do you know if there is a specific reason why it's not possible to validate 2 invoices with a different value in field "date"? 

If this is done to preserve invoicing sequence from job queue completion order, seems to me the check could be done on invoice_date field. Is there something I'm missing?

pedrobaeza commented:
I can't say, as I'm not using this module in newer versions (the speed improvements done on core have been enough for our customers).

Saran440 commented:
@francesco-ooops That's a good question. I'm not sure myself. I migrated from standard.
However, In my project I inherit this module for delete code validation too.

Should we delete it in this module because job queue processes tasks one by one?
@pedrobaeza

francesco-ooops commented:
@Saran440 I think the only thing we must make sure is not to validate together invoices with different invoice date as to preserve sequence

currently this check is done on field "date" which is not a visible field in the form and AFAIU is = creation date (not sure what is the purpose of this field)

adriapalleja commented:
@Saran440 @hbrunn How is the migration of this module going? Could we help somehow in this migration? Thanks!

Saran440 commented:
@adriapalleja Could you please review and approve this PR if it works? 
We can merge once 2 reviewers approve.

adriapalleja commented:
@Saran440 I've just reviewed and tested the module. Everything is working fine on my side, except from this topic in "Usage": 
- If any of the invoices is already enqueued, there will be a message saying so and avoiding to perform the process

The process is working as expected, but I would remove "there will be a message saying so" from the Usage Readme.

adriapalleja reviewed (changes requested):
Everything looking good! Only a minor change requested in the "Usage" Readme.

adriapalleja commented:
Hi @Saran440 ! Could you apply the minor changes suggested in order to move forward with this PR? Thanks!

francesco-ooops commented:
For the record, we edited this module internally by implementing the error message "You can't enqueue invoices with different dates." (seems to appear only in translation files at the moment) and forcing the user to validate invoices grouped by date (from older to newer in order to preserve sequence)

If date is not set in invoice, then "today" is used

Saran440 commented:
@francesco-ooops I tested and I think this version can use invoice with multiple dates. if you find any bugs, please explain and provide the steps to reproduce the error, and we will fix it.

For this module, it runs 1 queue per document, so the sequence will function correctly.
I have removed the validation error in this module because I couldn't find a reason for check.

@adriapalleja I updated readme, please recheck.

francesco-ooops commented:
@Saran440 if you're positive that enqueuing invoices with different dates results in invoice numbers following sequence correctly, then it's good for me 👍

sergiobstoj commented:
merge?

Saran440 commented:
@hbrunn Hi, Can you help merge, please?

hbrunn commented:
/ocabot merge nobump
